sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCast.scala

```diff
@@ -0,0 +1,155 @@
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.types._
+
+/**
+ * Unwrap casts in binary comparison operations with the following pattern:
+ *   `BinaryComparison(Cast(fromExp, _), Literal(value, toType))`
+ * This rule optimize expressions with this pattern by either replacing the cast with simpler
+ * constructs, or moving the cast from the expression side to the literal side, so they can be
+ * optimized away and pushed down to data sources.
+ *
+ * Currently this only handles the case where `fromType` (of `fromExp`) and `toType` are of
+ * integral types (i.e., byte, short, int and long). It checks to see if the literal `value` is
+ * within range (min, max) of the `fromType`. If this is true then it means we can safely cast the
+ * `value` to the `fromType` and thus able to move the cast to the literal side. Otherwise, it
+ * replaces the cast with different simpler constructs, such as
+ * `EqualTo(fromExp, Literal(max, fromType)` when input is
+ * `GreaterThanOrEqualTo(Cast(fromExp, fromType), Literal(max, toType))`
+ */
+object UnwrapCast extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case l: LogicalPlan => l transformExpressionsUp {
+      case e @ BinaryComparison(_, _) => unwrapCast(e)
+    }
+  }
+
+  private def unwrapCast(exp: Expression): Expression = exp match {
+    case BinaryComparison(Literal(_, _), Cast(_, _, _)) =>
+      def swap(e: Expression): Expression = e match {
+        case GreaterThan(left, right) => LessThan(right, left)
+        case GreaterThanOrEqual(left, right) => LessThanOrEqual(right, left)
+        case EqualTo(left, right) => EqualTo(right, left)
+        case EqualNullSafe(left, right) => EqualNullSafe(right, left)
+        case LessThanOrEqual(left, right) => GreaterThanOrEqual(right, left)
+        case LessThan(left, right) => GreaterThan(right, left)
+        case other => other
+      }
+      swap(unwrapCast(swap(exp)))
+
+    case BinaryComparison(Cast(fromExp, _, _), Literal(value, toType)) =>
+      val fromType = fromExp.dataType
+      if (!fromType.isInstanceOf[IntegralType] || !toType.isInstanceOf[IntegralType]
+        || !Cast.canUpCast(fromType, toType)) {
+        return exp
+      }
+
+      // Check if the literal value is within the range of the `fromType`, and handle the boundary
+      // cases in the following
+      val toIntegralType = toType.asInstanceOf[IntegralType]
+      val (min, max) = getRange(fromType)
+      val (minInToType, maxInToType) =
+        (Cast(Literal(min), toType).eval(), Cast(Literal(max), toType).eval())
+
+      // Compare upper bounds
+      val maxCmp = toIntegralType.ordering.asInstanceOf[Ordering[Any]].compare(value, maxInToType)
+      if (maxCmp > 0) {
+        exp match {
+          case EqualTo(_, _) | GreaterThan(_, _) | GreaterThanOrEqual(_, _) =>
```

Comment:
`case EqualTo(_, _) | GreaterThan(_, _) | GreaterThanOrEqual(_, _) if maxCmp > 0 ` might be more readable, and you can combine the next pattern matching.

Reply:
Do you mean combining the cases like:
```scala
exp match {
case <case1> if maxCmp > 0 =>
case <case2> if maxCmp == 0 =>
case <case3> if minCmp < 0 =>
case <case4> if minCmp == 0 =>
...
```
?. This will end up having many cases since they do not fully overlap with each other.

Reply:
I was thinking about that. Let's revisit it later.